server/test/firestore_test.dart

```diff
@@ -0,0 +1,187 @@
+import 'package:aqueduct/aqueduct.dart';
+import 'package:aqueduct_test/aqueduct_test.dart';
+import 'package:fixnum/fixnum.dart';
+import 'package:googleapis/firestore/v1.dart' as firestore;
+import 'package:googleapis/discovery/v1.dart';
+import 'package:http/http.dart' as http;
+import 'package:test/test.dart';
+
+import 'package:server/proto/rule.pb.dart';
+import 'package:server/service/firestore.dart';
+import 'package:server/utils.dart';
+
+void main() {
+  const mockServerPort = 8002;
+  final mockFirestoreServer = MockHTTPServer(mockServerPort);
+  const url = 'http://localhost:$mockServerPort/';
+
+  const projectId = 'spreadsheet-dv360-plugin';
+  const databaseId = '(default)';
+  const ruleName = 'testRule';
+  const userId = '1234567890';
+  const parent = 'projects/$projectId/databases/$databaseId/documents';
+  const ruleResourceName =
+      '$parent/${FirestoreClient.usersName}/$userId/'
+      '${FirestoreClient.rulesName}';
+  const userCollectionResourceName =
+      '$parent/${FirestoreClient.usersName}';
+  const encryptedRefreshToken = 'abc123';
+
+  final rule = Rule()
+    ..name = ruleName
+    ..action = (Action()
+      ..type = Action_Type.CHANGE_LINE_ITEM_STATUS
+      ..changeLineItemStatusParams = (ChangeLineItemStatusParams()
+        ..lineItemIds.add(Int64(12345))
+        ..advertiserId = Int64(67890)
+        ..status = ChangeLineItemStatusParams_Status.PAUSED))
+    ..schedule = (Schedule()
+      ..type = Schedule_Type.REPEATING
+      ..timezone = 'America/Los_Angeles'
+      ..repeatingParams =
+          (Schedule_RepeatingParams()..cronExpression = '* * * * *'));
+
+  final client = http.Client();
+  final firestoreClient = FirestoreClient(client, projectId, databaseId, url);
+
+  Request request;
+  String returnedString;
+  firestore.Document userDocument;
+
+  setUpAll(() async {
+    await mockFirestoreServer.open();
+  });
+
+  tearDownAll(() async {
+    await mockFirestoreServer.close();
+  });
+
+  tearDown(() async {
+    mockFirestoreServer.clear();
+  });
+
+  group('Success case: createRule()', () {
+    setUp(() async {
+      final ruleAsDocument = rule.toDocument();
+      ruleAsDocument.name = ruleName;
```

Comment:
does this already set in `rule.toDocument()`?

Reply:
Oops, that line should be `ruleAsDocument.name = documentName`. I'll fix it 😅